### PR TITLE
Fixes #3516: more override functionalities for plugins

### LINKED ISSUE #3516: Allow dynamic containment configuration for plugins
### Description
Currently the plugins system do not allow overrides that:
 * change the priority of existing containers
 * add new containers

Those would be nice additions.

### PR DESCRIPTION
**Pull request by mbarto**
## Description
Allow more override to work on plugins:
 * override of containers priority
 * adding a new override block for a custom container, so that plugins can be included in every container, by configuration

Some cleanup as a bonus.

## Issues
 - Fix #3516

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)

 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)
You cannot override plugin container priorities, neither configure a new container for a plugin.

**What is the new behavior?**
You can override plugin container priorities, and configure a new container for a plugin.

**Does this PR introduce a breaking change?** (check one with "x", remove the other)

 - [ ] Yes
 - [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: web/client/utils/PluginsUtils.js [hunk 1]
```diff
@@ -273,20 +337,24 @@ const PluginsUtils = {
...
-            const Plugin = loadedPlugins && loadedPlugins[pluginDef.name] || (pluginDef.plugin && !pluginDef.plugin.loadPlugin && pluginDef.plugin);
+            const impl = loadedPlugins[pluginDef.name] ||
+                !pluginDef.plugin.loadPlugin && pluginDef.plugin;
+            const id = isObject(pluginDef) ? pluginDef.id : null;
+            const stateSelector = isObject(pluginDef) ? pluginDef.stateSelector : id || undefined;
+            const Plugin = getPluginImplementation(impl, stateSelector);
             const result = (props) => {
                 return Plugin ? (<Plugin key={pluginDef.id}
-                    {...props} {...pluginDef.cfg} pluginCfg={pluginDef.cfg} />) : loaderComponent;
+                    {...props} {...pluginDef.cfg} pluginCfg={pluginDef.cfg} items={pluginDef.items || []}/>) : loaderComponent;
```
Comment (allyoucanmap) [inline web/client/utils/PluginsUtils.js]: while testing I noticed here a nested configuration
eg: 
```
{
    "name": "TOC",
    "priority": 1,
    "glyph": "1-layer",
    "position": 2,
    "size": 256,
    "tooltip": "Layers",
    "container": "left-menu",
    "cfg": {
        "name": "TOC",
        "cfg": {
            "activateMapTitle": false,
            "activateRemoveLayer": false,
            "activateSettingsTool": false,
            "activateTitleTooltip": false
        }
    },
    "items": []
}
```
so these props from cfg will not apply to Plugin `{...pluginDef.cfg} pluginCfg={pluginDef.cfg}`

File: web/client/utils/PluginsUtils.js [hunk 2]
```diff
@@ -273,20 +337,24 @@ const PluginsUtils = {
...
-            impl: includeLoaded(name, loadedPlugins, impl.loadPlugin || impl.displayName || impl.prototype.isReactComponent ? impl : impl(stateSelector)),
+            impl: includeLoaded(name, loadedPlugins, getPluginImplementation(impl, stateSelector)),
             cfg: assign({}, impl.cfg || {}, isObject(pluginDef) ? parsePluginConfig(state, plugins.requires, pluginDef.cfg) : {}),
             items: getPluginItems(state, plugins, pluginsConfig, name, id, isDefault, loadedPlugins)
         };
     },
     getPluginItems,
-    getConfiguredPlugin: (pluginDef, loadedPlugins, loaderComponent) => {
+    getConfiguredPlugin: (pluginDef, loadedPlugins = {}, loaderComponent) => {
         if (pluginDef) {
```
Comment (allyoucanmap) [inline web/client/utils/PluginsUtils.js]: if plugin is initialized with a string in localConfig.json the cfg in pluginDef contains chars of the name
```
{
    "name": "Search",
    "position": 1,
    "priority": 1,
    "cfg": {
        "0": "S",
        "1": "e",
        "2": "a",
        "3": "r",
        "4": "c",
        "5": "h"
    },
    "items": []
}
```

File: web/client/utils/PluginsUtils.js (hunk 1 above)
Comment (mbarto) [reply web/client/utils/PluginsUtils.js]: @allyoucanmap Good catch, I committed a fix for this

File: web/client/utils/PluginsUtils.js (hunk 2 above)
Comment (mbarto) [reply web/client/utils/PluginsUtils.js]: @allyoucanmap the previous fix should also fix this one

File: web/client/plugins/__tests__/pluginsTestUtils.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import { combineEpics, createEpicMiddleware } from 'redux-observable';
+
+const createRegisterActionsMiddleware = (actions) => {
+    return () => next => action => {
+        actions.push(action);
+        return next(action);
+    };
+};
+
+export const getPluginForTest = (pluginDef, storeState, plugins) => {
```
Comment (offtherailz) [inline web/client/plugins/__tests__/pluginsTestUtils.js]: Looks a very good utility. Can you jsdoc them?

Comment (mbarto) [reply web/client/plugins/__tests__/pluginsTestUtils.js]: it's going to be documented in the dev plugins documentation

**Pull request merged**